README.md

```diff
@@ -159,15 +159,22 @@ any errors encountered during the request.
 
 ## Body
 
-If the server's response contains a body it should follow the requirements for [GraphQL response](https://graphql.github.io/graphql-spec/June2018/#sec-Response).
+If the server's response contains a body it MUST follow the requirements for [GraphQL response](https://graphql.github.io/graphql-spec/June2018/#sec-Response).
 
-Note: For any non-2XX response, the client should not rely on the body to be in GraphQL format since the source of the response
-may not be the GraphQL server but instead some intermediary such as API gateways, firewalls, etc.
+Note: For any non-2XX response, the client should not rely on the body to be in GraphQL format.
+The source of the response may not be the GraphQL server, but instead some intermediary such as API gateways, firewalls, etc.
 
 ## Status Codes
 
-If the `data` entry in the response has any value other than `null` (when the operation has successfully executed
-without error) then the response should use the 200 (OK) status code.
-If the operation failed before or during execution, due to a syntax error, missing information, validation error
-or any other reason, then response should use the 4XX or 5XX status codes.
+A server must differentiate between different error conditions.
+
+The result of executing a GraphQL operation may contain partial data as well as encountered errors.
+Errors that happen during execution of the GraphQL operation become part of the result.
+If the server successfully executed parts of the operation and returns a well-formed response
+that contains the key `data`, the status code MUST be `200` (OK).
```

Comment:
I am not sure yet on how we should handle the case where `data == null`. Here is a relevant section of the GraphQL spec:

> If an error was encountered before execution begins, the data entry should not be present in the result.
>
> If an error was encountered during the execution that prevented a valid response, the data entry in the response should be null.

What would be examples of errors during execution that prevent a valid response?

Reply:
Response serialization errors.

Reply:
If there were an non-null root field that threw an error.

Reply:
Do you think it would be appropriate to return a non-200 status code for those cases?

Reply:
For my one, no. But the reference implementation does, I think? @IvanGoncharov  is on the fence about it, I think? (Sorry for all the "I think"s!)

Reply:
Response serialization errors, e.g. happens after GraphQL resolves but for some reason cannot write the result out, should be a 500 to my mind.

Reply:
@benjie i agree, errors on non-null field are still within the realm of GraphQL execution. The behaviour is consistent with what happens within nested fields. The response is still well-formed GraphQL, including appropriate errors.

Returning a different status code would be the kind of special casing that we generally want to avoid.